bgpd/bgp_bmp.c

```diff
@@ -271,29 +312,43 @@ static void bmp_per_peer_hdr(struct stream *s, struct peer *peer,
 	stream_putc(s, flags);
 
 	/* Peer Distinguisher */
-	memset (&peer_distinguisher[0], 0, 8);
-	stream_put(s, &peer_distinguisher[0], 8);
+	stream_put(s, (uint8_t *)&peer_distinguisher, 8);
 
 	/* Peer Address */
-	if (peer->su.sa.sa_family == AF_INET6)
-		stream_put(s, &peer->su.sin6.sin6_addr, 16);
-	else if (peer->su.sa.sa_family == AF_INET) {
+	/* Set to 0 if it's a LOC-RIB INSTANCE (RFC 9069) or if it's not an
+	 * IPv4/6 address
+	 */
+	if (is_locrib || (peer->su.sa.sa_family != AF_INET6 &&
+			  peer->su.sa.sa_family != AF_INET)) {
 		stream_putl(s, 0);
 		stream_putl(s, 0);
 		stream_putl(s, 0);
-		stream_put_in_addr(s, &peer->su.sin.sin_addr);
-	} else {
 		stream_putl(s, 0);
+	} else if (peer->su.sa.sa_family == AF_INET6)
+		stream_put(s, &peer->su.sin6.sin6_addr, 16);
+	else if (peer->su.sa.sa_family == AF_INET) {
 		stream_putl(s, 0);
 		stream_putl(s, 0);
 		stream_putl(s, 0);
+		stream_put_in_addr(s, &peer->su.sin.sin_addr);
 	}
 
 	/* Peer AS */
-	stream_putl(s, peer->as);
+	/* set peer ASN but for LOC-RIB INSTANCE (RFC 9069) put the local bgp
+	 * ASN if available or 0
+	 */
+	as_t asn = !is_locrib ? peer->as : bgp ? bgp->as : 0L;
```

Comment:
the comment above the line explains why, maybe not in depth enough...


In case of `!is_locrib` aka *adj-rib-in*:
[RFC7854 BMP (Adj-In-Pre/Post)](https://datatracker.ietf.org/doc/html/rfc7854#section-4.2)
  ```
o  Peer AS: The Autonomous System number of the peer from which the
      encapsulated PDU was received. 
```
so `peer->as` if i'm not mistaken


In case of `is_locrib` :
[RFC9069 BMP Local-RIB](https://datatracker.ietf.org/doc/html/rfc9069#section-5.1)
```
Peer Autonomous System (AS):
    Set to the primary router BGP autonomous system number (ASN).
```
so `bgp->as` if i'm not mistaken too


the case of `bgp` being NULL or not is just extra safety to avoid unfortunate null pointer dereferences.

is something going over my head about the RFCs or the `struct peer` / `struct bgp` fields?
ok for rewriting if you really want it but 2 ternaries do not look that bad to me. please confirm again that you want to change this to if else statements.